### PR TITLE
[#477][part-0] feat: add ShuffleManagerServer impl

### PR DESCRIPTION
**Pull request by advancedxy**
### What changes were proposed in this pull request?
1. add shuffle manager proto
2. the corresponding client and server impls
3. add the RssShuffleManagerInterface 

### Why are the changes needed?
This is the first part of #477. 
To support re-submit spark stage, the ShuffleManagerServer should be introduced first.

### Does this PR introduce _any_ user-facing change?
No.

### How was this patch tested?
Added UTs and one integration test.

### CODE REVIEW

File: common/src/main/java/org/apache/uniffle/common/metrics/DummyGRPCMetrics.java [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.uniffle.common.metrics;
+
+public class DummyGRPCMetrics extends GRPCMetrics {
```
Comment (jerqi) [inline common/src/main/java/org/apache/uniffle/common/metrics/DummyGRPCMetrics.java:20]: Em.. I  feel that this class is used for test when I see this class. What is this class used for? Could we give it a better name or add some comments? I guess that `EmptyGRPCMetrics` may be fit.

File: integration-test/spark-common/src/test/java/org/apache/uniffle/test/SimpleShuffleServerManagerE2ETest.java [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+
+import org.apache.uniffle.client.request.RssReportShuffleFetchFailureRequest;
+import org.apache.uniffle.client.response.RssReportShuffleFetchFailureResponse;
+import org.apache.uniffle.common.rpc.StatusCode;
+import org.apache.uniffle.shuffle.manager.DummyRssShuffleManager;
+import org.apache.uniffle.shuffle.manager.RssShuffleManagerInterface;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+public class SimpleShuffleServerManagerE2ETest extends ShuffleServerManagerTestBase {
```
Comment (jerqi) [inline integration-test/spark-common/src/test/java/org/apache/uniffle/test/SimpleShuffleServerManagerE2ETest.java:30]: Why is test is E2E test? What's the difference from integration tests? Will we introduce the concept of `E2E` tests?

File: internal-client/src/main/java/org/apache/uniffle/client/response/RssReportShuffleFetchFailureResponse.java [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+    return builder
+        .setStatus(getStatusCode().toProto())
+        .setMsg(getMessage())
+        .setReSubmitWholeStage(reSubmitWholeStage)
+        .build();
+  }
+
+  public static RssReportShuffleFetchFailureResponse fromProto(ReportShuffleFetchFailureResponse response) {
+    return new RssReportShuffleFetchFailureResponse(
+        // todo: add fromProto for StatusCode
```
Comment (jerqi) [inline internal-client/src/main/java/org/apache/uniffle/client/response/RssReportShuffleFetchFailureResponse.java:46]: If we add `todo`, we would better add an issue to track it.

File: integration-test/spark-common/src/test/java/org/apache/uniffle/test/SimpleShuffleServerManagerE2ETest.java (hunk 2 above)
Comment (jerqi) [reply integration-test/spark-common/src/test/java/org/apache/uniffle/test/SimpleShuffleServerManagerE2ETest.java:30]: If they have no difference, I prefer using `IntegrationTest` here to unify our style.

File: internal-client/src/main/java/org/apache/uniffle/client/response/RssReportShuffleFetchFailureResponse.java (hunk 3 above)
Comment (jerqi) [reply internal-client/src/main/java/org/apache/uniffle/client/response/RssReportShuffleFetchFailureResponse.java:46]: Could we add ISSUE ID in the comment?

**Review by jerqi — APPROVED** [review]
LGTM

**Review by jerqi — DISMISSED** [review]

File: client-spark/common/src/main/java/org/apache/uniffle/shuffle/manager/ShuffleManagerGrpcService.java [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+      LOG.warn(msg);
+      code = RssProtos.StatusCode.INVALID_REQUEST;
+      reSubmitWholeStage = false;
+    } else {
+      RssShuffleStatus status = shuffleStatus.computeIfAbsent(request.getShuffleId(), key -> {
+            int partitionNum = shuffleManager.getPartitionNum(key);
+            return new RssShuffleStatus(partitionNum, stageAttempt);
+          }
+      );
+      int c = status.resetStageAttemptIfNecessary(stageAttempt);
```
Comment (zuston) [inline client-spark/common/src/main/java/org/apache/uniffle/shuffle/manager/ShuffleManagerGrpcService.java:63]: what's meaning of `c` ?

File: client-spark/common/src/main/java/org/apache/uniffle/shuffle/manager/ShuffleManagerServerFactory.java [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+
+  public ShuffleManagerServerFactory(RssShuffleManagerInterface shuffleManager, RssConf conf) {
+    this.shuffleManager = shuffleManager;
+    this.conf = new RssBaseConf();
+    this.conf.addAll(conf);
+  }
+
+  public GrpcServer getServer() {
+    String type = conf.getString(RssBaseConf.RPC_SERVER_TYPE);
+    if (type.equals(ServerType.GRPC.name())) {
```
Comment (zuston) [inline client-spark/common/src/main/java/org/apache/uniffle/shuffle/manager/ShuffleManagerServerFactory.java:37]: I prefer `ServerType.GRPC.name().equals(type)`

File: client-spark/common/src/main/java/org/apache/uniffle/shuffle/manager/ShuffleManagerServerFactory.java [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+  private final RssBaseConf conf;
+
+  public ShuffleManagerServerFactory(RssShuffleManagerInterface shuffleManager, RssConf conf) {
+    this.shuffleManager = shuffleManager;
+    this.conf = new RssBaseConf();
+    this.conf.addAll(conf);
+  }
+
+  public GrpcServer getServer() {
+    String type = conf.getString(RssBaseConf.RPC_SERVER_TYPE);
```
Comment (zuston) [inline client-spark/common/src/main/java/org/apache/uniffle/shuffle/manager/ShuffleManagerServerFactory.java:36]: It looks `RPC_SERVER_TYPE` should be refactored as `ENUM` type

File: internal-client/src/main/java/org/apache/uniffle/client/factory/ShuffleManagerClientFactory.java [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+
+  public static ShuffleManagerClientFactory getInstance() {
+    return LazyHolder.INSTANCE;
+  }
+
+  private ShuffleManagerClientFactory() {
+  }
+
+  public ShuffleManagerGrpcClient createShuffleManagerClient(ClientType clientType, String host, int port) {
+    if (clientType.equals(ClientType.GRPC)) {
```
Comment (zuston) [inline internal-client/src/main/java/org/apache/uniffle/client/factory/ShuffleManagerClientFactory.java:37]: ditto.

File: internal-client/src/main/java/org/apache/uniffle/client/impl/grpc/ShuffleManagerGrpcClient.java [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+import org.apache.uniffle.client.response.RssReportShuffleFetchFailureResponse;
+import org.apache.uniffle.common.exception.RssException;
+import org.apache.uniffle.proto.RssProtos.ReportShuffleFetchFailureRequest;
+import org.apache.uniffle.proto.RssProtos.ReportShuffleFetchFailureResponse;
+import org.apache.uniffle.proto.ShuffleManagerGrpc;
+
+public class ShuffleManagerGrpcClient extends GrpcClient implements ShuffleManagerClient {
+
+  private static final Logger LOG = LoggerFactory.getLogger(ShuffleManagerGrpcClient.class);
+  private static final long RPC_TIMEOUT_DEFAULT_MS = 60000;
```
Comment (zuston) [inline internal-client/src/main/java/org/apache/uniffle/client/impl/grpc/ShuffleManagerGrpcClient.java:36]: Should be configurable?

**Review by zuston — DISMISSED** [review]
LGTM. Left minor comments

**Review by zuston — APPROVED** [review]

**Pull request merged**